### PR TITLE
Store secret keys in separate files [ECR-2518]

### PR DESCRIPTION
**Pull request by mvlabat**
## Overview

This PR changes the way we store secret keys for our nodes. CLI has been updated for these changes as well.

---

See:
- https://jira.bf.local/browse/ECR-2518
- https://jira.bf.local/browse/ECR-2519

### Definition of Done

- [ ] There are no TODOs left in the merged code
- [ ] Change is covered by automated tests
- [ ] Benchmark results are attached (if applicable)
- [ ] The [coding guidelines] are followed
- [ ] Public API has proper documentation
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

[coding guidelines]: https://github.com/exonum/exonum/blob/master/CONTRIBUTING.md#conventions

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -109,7 +113,7 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 
 #### exonum-crypto
 
-- Added `utils` module with functions `create_keys_file` for creating
+- Added `utils` module with functions `generate_keys_file` for creating
```
Comment (aleksuss) [inline CHANGELOG.md]: 👍

File: exonum/src/helpers/fabric/details.rs [hunk 2]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
...
+                let stdin_locked = stdin.lock();
+                Self::prompt_passphrase(stdin_locked, io::stdout()).ok()
+            }).unwrap()
+            .into_bytes()
+    }
+
+    fn prompt_passphrase<R: BufRead, W: Write>(mut reader: R, mut writer: W) -> io::Result<String> {
+        loop {
+            write!(&mut writer, "Enter passphrase (empty for no passphrase): ").unwrap();
+            let password = rpassword::read_password_with_reader(Some(&mut reader))?;
```
Comment (ilammy) [inline exonum/src/helpers/fabric/details.rs]: I'd suggest using `rpassword::read_password_from_tty` instead of explicitly manipulating stdin and stdout.
It is more correct to read passwords directly from terminal in UNIX world.

I guess you have used `BufRead` exactly to be able to write unit-tests for this function. However, I'm not sure that we need to test this behavior, retrying not that complex and we can trust `rpassword` to do its thing correctly. This test seems to provide little value. It would be more useful to have an integration test which verifies environment variable handling together with that `NO_PASSWORD` thing.

Also, I'm curious: how does this behave if there is no interactive terminal? E.g., when Exonum node is launched as a service with stdin and stdout bound to `/dev/null`?

File: exonum/src/helpers/fabric/details.rs [hunk 3]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
...
+            if password == rpassword::read_password_with_reader(Some(&mut reader))? {
+                return Ok(password);
+            }
+            writeln!(&mut writer, "Passphrases do not match.  Try again.").unwrap();
+        }
+    }
+
+    fn create_secret_key_file(
+        secret_key_path: impl AsRef<Path>,
+        passphrase: impl AsRef<[u8]>,
```
Comment (ilammy) [inline exonum/src/helpers/fabric/details.rs]: I'd suggest calling `get_passphrase` from `create_secret_key_file` directly instead of passing the passphrase in. These is not need to use a passphrase outside of this function, and the function itself is an internal utility rather than general-purpose public function. This should be a nicer API for where it's actually used.

I'm not sure whether passing EXONUM_SERVICE_PASS here will read better though...

File: exonum/src/helpers/fabric/details.rs [hunk 4]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
 
         (external_address, listen_address)
     }
+
+    /// * `key` - expects EXONUM_SERVICE_PASS or EXONUM_CONSENSUS_PASS.
+    fn get_passphrase(context: &Context, key: &str) -> Vec<u8> {
```
Comment (ilammy) [inline exonum/src/helpers/fabric/details.rs]: I'm a bit concerned that someone may reuse `get_passphrase` somewhere later to read an existing passphase, thinking that it's a general passphrase-reading function. Currently it is geared towards this only use: reading a fresh new consensus or service password for key files. Maybe we should consider giving this function more unwieldy name like `get_new_exonum_key_file_passphrase` or something alike so that it stays an internal utility.

File: exonum/src/helpers/fabric/details.rs [hunk 5]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
...
+        }
+    }
+
+    fn create_secret_key_file(
+        secret_key_path: impl AsRef<Path>,
+        passphrase: impl AsRef<[u8]>,
+    ) -> PublicKey {
+        let secret_key_path = secret_key_path.as_ref();
+        if secret_key_path.exists() {
+            panic!("{}: File exists", secret_key_path.to_string_lossy(),);
```
Comment (ilammy) [inline exonum/src/helpers/fabric/details.rs]: These panics and unwraps make me cringe a bit. We can be more nice and return a Result from this function and _then_ panic in one place if we wish to. A function like this should not make decisions on error handling.

I'm not against leaving it as is (because of gazillion of other unwraps that are already present in this module), but following the boy scout rule is good for one's karma.

File: exonum/src/helpers/fabric/details.rs [hunk 6]
```diff
@@ -459,8 +545,24 @@ impl Command for GenerateNodeConfig {
...
+        let service_public_key =
+            Self::create_secret_key_file(&service_secret_key_path, &service_passphrase);
+
+        let pub_config_folder = Path::new(&pub_config_path).parent().unwrap();
+        let consensus_secret_key = if consensus_secret_key_path.is_absolute() {
+            consensus_secret_key_path
+        } else {
+            path_relative_from(&consensus_secret_key_path, &pub_config_folder).unwrap()
+        };
+        let service_secret_key = if service_secret_key_path.is_absolute() {
```
Comment (ilammy) [inline exonum/src/helpers/fabric/details.rs]: It feels that it would be better to have this `is_absolute` branch as a part of `path_relative_from`.

I also believe that it will be more nice to read if the arguments switch places:

```rust
path_relative_from(&pub_config_folder, service_secret_key_path)
```

File: exonum/src/helpers/fabric/mod.rs [hunk 7]
```diff
@@ -79,9 +79,35 @@ pub struct Argument {
     pub required: bool,
     /// Help message.
     pub help: &'static str,
+    /// Used for named arguments: defines whether the argument is a flag or not (false for flags).
+    pub takes_value: bool,
 }
 
 impl Argument {
+    /// Creates a new falg with `long` and optionally `short` names.
```
Comment (ilammy) [inline exonum/src/helpers/fabric/mod.rs]:
```suggestion
    /// Creates a new flag with `long` and optionally `short` names.
```

File: exonum/src/helpers/mod.rs [hunk 8]
```diff
@@ -77,3 +79,62 @@ pub fn generate_testnet_config(count: u16, start_port: u16) -> Vec<NodeConfig> {
...
+#[test]
+fn test_path_relative_from() {
+    let baz: PathBuf = "/foo/bar/baz".into();
+    let bar: PathBuf = "/foo/bar".into();
+    let quux: PathBuf = "/foo/bar/quux".into();
+    assert_eq!(path_relative_from(&bar, &baz), Some("../".into()));
+    assert_eq!(path_relative_from(&baz, &bar), Some("baz".into()));
+    assert_eq!(path_relative_from(&quux, &baz), Some("../quux".into()));
+    assert_eq!(path_relative_from(&baz, &quux), Some("../baz".into()));
+    assert_eq!(path_relative_from(&bar, &quux), Some("../".into()));
```
Comment (ilammy) [inline exonum/src/helpers/mod.rs]: I'd like to see more edge cases in tests for a utility function which may be reused in surprising places later.

- relative paths (as both arguments)
- special path components (`.` and `..`)
- trying to escape the root directory (`/foo/../../../../../../../../../../../bar`)
- any other stupid things that you may think of (e.g., `("/", "/")`, referring the directory itself, etc.)

File: exonum/src/node/mod.rs [hunk 9]
```diff
@@ -268,6 +269,51 @@ pub struct NodeConfig {
...
+impl NodeConfig<PathBuf> {
+    /// Converts `NodeConfig<PathBuf>` to `NodeConfig<SecretKey>` reading the key files.
+    pub fn read_secret_keys(
+        self,
+        config_file_path: impl AsRef<Path>,
+        consensus_passphrase: &[u8],
+        service_passphrase: &[u8],
+    ) -> NodeConfig {
+        let config_folder = config_file_path.as_ref().parent().unwrap();
+        let consensus_key_path = if self.consensus_secret_key.is_absolute() {
```
Comment (ilammy) [inline exonum/src/node/mod.rs]: I'd suggest to extract this branch into a utility function.

File: exonum/tests/config.rs [hunk 10]
```diff
@@ -32,14 +32,17 @@ use exonum::{
...
-    fs::{File, OpenOptions},
+    fs::{self, File, OpenOptions},
     io::{Read, Write},
     panic,
-    path::Path,
+    path::{Path, PathBuf},
 };
 
+const EXONUM_CONSENSUS_PASS: &str = "EXONUM_CONSENSUS_PASS";
+const EXONUM_SERVICE_PASS: &str = "EXONUM_SERVICE_PASS";
```
Comment (ilammy) [inline exonum/tests/config.rs]: Can't we somehow export these constants for tests instead of duplicating them? Or it looks too ugly?

File: exonum/src/helpers/fabric/mod.rs (hunk 7 above)
Comment (vitvakatu) [inline exonum/src/helpers/fabric/mod.rs]:
```suggestion
    /// Creates a new flag with `long` and optionally `short` names.
```

File: exonum/src/helpers/fabric/details.rs [hunk 11]
```diff
@@ -416,6 +471,29 @@ impl Command for GenerateNodeConfig {
                 "listen-address",
                 false,
             ),
+            Argument::new_named(
+                CONSENSUS_KEY_PATH,
+                false,
+                "Path to the pem file storing consensus private key (default: ./consensus.toml)",
```
Comment (aleksuss) [inline exonum/src/helpers/fabric/details.rs]: `pem` is redundant here.

File: exonum/tests/config.rs (hunk 10 above)
Comment (mvlabat) [reply exonum/tests/config.rs]: I looked through the Exonum code and found out that we do the same for some other constants (I don't remember which ones exactly). If I make them public, I'll have to reexport them in several modules and write documentation... Meh )

File: exonum/Cargo.toml [hunk 12]
```diff
@@ -53,6 +53,7 @@ uuid = { version = "=0.7.1", features = ["serde"] }
 snow = "=0.4.0"
 rust_decimal = "=0.10.2"
 protobuf = { version = "2.2.0", features = ["with-serde"] }
+rpassword = "2.0.0"
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum/Cargo.toml]: @ilammy why?

File: exonum/src/helpers/fabric/details.rs (hunk 6 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: I didn't want to edit the function which once was in the `stdlib`. I'd suggest to leave it as-is. It was somehow tested and reliable, I fell like it's better to keep that expected original behavior.

File: exonum/src/helpers/fabric/details.rs (hunk 4 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: It's a private method for this command.. Why would someone reuse it?
I'd rather not use such a verbose name, though I can add some comments in order to make it clear what this function does.

File: exonum/src/helpers/fabric/details.rs (hunk 3 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: Yeah.. I didn't want to pass variable name to this function, that why I kept it this way. Really dunno which way is really better

File: exonum/Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (aleksuss) [reply exonum/Cargo.toml]: @ilammy we pin to a latest version if version less than 1.0.

File: exonum/src/helpers/fabric/details.rs (hunk 2 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: > This test seems to provide little value.

I think it's important to test our prompt behavior: what exactly is displayed to used and under what conditions. I can't see a way to do it in our integration tests.

> Also, I'm curious: how does this behave if there is no interactive terminal? E.g., when Exonum node is launched as a service with stdin and stdout bound to `/dev/null`?

You can always pass the password with environment variables, so you won't be prompted to enter the password manually.

File: exonum/src/helpers/fabric/details.rs (hunk 4 above)
Comment (aleksuss) [reply exonum/src/helpers/fabric/details.rs]: @mvlabat To hold a passphrase in `Vec` it's an unsecured approach. For that propose you can use https://docs.rs/smallvec/0.6.6/smallvec/struct.SmallVec.html.

File: exonum/src/node/mod.rs [hunk 13]
```diff
@@ -268,6 +269,51 @@ pub struct NodeConfig {
...
+            config_folder.join(&self.consensus_secret_key)
+        };
+        let service_key_path = if self.service_secret_key.is_absolute() {
+            self.service_secret_key
+        } else {
+            config_folder.join(&self.service_secret_key)
+        };
+
+        let consensus_secret_key = read_keys_from_file(&consensus_key_path, consensus_passphrase)
+            .expect("Could not read consensus_secret_key from file")
```
[the remark this answers is not in the document]
Comment (mvlabat) [reply exonum/src/node/mod.rs]: It's always appended to the `expect` message. See
https://github.com/rust-lang/rust/blob/master/src/libcore/result.rs#L835
https://github.com/rust-lang/rust/blob/master/src/libcore/result.rs#L1008

File: exonum/src/helpers/fabric/details.rs (hunk 4 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: Saw the same comment for your PR and forgot about it.. Shame! )

File: exonum/src/helpers/fabric/details.rs (hunk 6 above)
Comment (ilammy) [reply exonum/src/helpers/fabric/details.rs]: If it were still around in stdlib as a deprecated function then I'd suggest using it instead. However, it was removed then I don't really buy the sentimental argument of keeping all the old quirks in naming, behavior, and argument order when we can do better now. It's not like new developers looking at this code will remember the old behavior. Plus, we have our own tests to verify that this function behaves as we expect it to.

Though, argument naming and order is a subjective matter so I'm most certainly not blocking this PR if you believe that we should keep the original naming and argument order.

File: exonum/tests/config.rs (hunk 10 above)
Comment (ilammy) [reply exonum/tests/config.rs]: I'm not talking about making a really 'public' public export, with docs and stuff. Something like this in the Exonum module

```rust
#[cfg(test)]
pub use module::of::the::EXONUM_SERVICE_PASS;
```

Not sure if it will work though. If it's too hard or too ugly to read then forget it.

File: exonum/Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ilammy) [reply exonum/Cargo.toml]: Hm... that's news to me. I thought we pin _everything_. I guess that's a false impression made by 99% of Rust stuff still having 0.x.x versions :rofl: 

However, I also see some 1.0.0+ crates being pinned, such as `os_info`, `protobuf`. As well as some pre-1.0.0 stuff not pinned, such as `failure`, `erased_serde`.

I guess we need to spell this rule out in CONTRIBUTING.md or somewhere if that's our strategy on dependency management.

[the remark this answers is not in the document]
Comment (aleksuss) [reply exonum/Cargo.toml]: Be sure we will do it

File: exonum/src/helpers/fabric/details.rs (hunk 6 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: Yeah, I haven't brought up the main argument and I'm talking not against renaming the function right now.. I don't want to put the branch there. It really has nothing to do with the purpose of that function. The logic is to keep absolute paths as-is, because that's what user, I think, expects to see in the config if they pass an absolute path to the command line.

File: exonum/src/helpers/fabric/details.rs (hunk 4 above)
Comment (ilammy) [reply exonum/src/helpers/fabric/details.rs]: I'm not sure how `SmallVec` helps there. Isn't it (slightly) insecure because we should securely zero out memory before dropping it?

File: exonum/src/helpers/fabric/details.rs (hunk 2 above)
Comment (ilammy) [reply exonum/src/helpers/fabric/details.rs]: > You can always pass the password with environment variables, so you won't be prompted to enter the password manually.

But what if I forget to set the environment variable? Will Exonum node hang indefinitely, assume an empty password, or prompt the user on some random terminal for the password?

File: exonum/src/helpers/fabric/builder.rs [hunk 14]
```diff
@@ -139,6 +136,31 @@ impl NodeBuilder {
...
+            .get(keys::NODE_CONFIG)
+            .expect("could not find node_config");
+        let db = Run::db_helper(ctx, &config.database);
+        let services: Vec<Box<dyn Service>> = self
+            .service_factories
+            .into_iter()
+            .map(|mut factory| factory.make_service(ctx))
+            .collect();
+
+        let consensus_passphrase = env::var(EXONUM_CONSENSUS_PASS).unwrap_or_default();
```
Comment (slowli) [inline exonum/src/helpers/fabric/builder.rs]: Am I reading this correctly that the password can only be input from env when starting a node? This doesn't sound right.

File: exonum/src/helpers/fabric/details.rs [hunk 15]
```diff
@@ -416,6 +471,29 @@ impl Command for GenerateNodeConfig {
...
+            Argument::new_named(
+                SERVICE_KEY_PATH,
+                false,
+                "Path to the file storing service private key (default: ./service.toml)",
+                "s",
+                "service-path",
+                false,
+            ),
+            Argument::new_flag(
+                NO_PASSWORD,
```
Comment (slowli) [inline exonum/src/helpers/fabric/details.rs]: What is the utility of this flag? I'm not an expert in designing secure UX, but I think that the *only* case where we might not ask for a password is creating a test network. We would want to provide domain separation (i.e., make it impossible to run a test network node in the "real" network, or vice versa; this might include different address format if/when address appear in Exonum) and otherwise make clear that a test network is not secure and should not be used in production.

File: exonum/src/helpers/fabric/details.rs [hunk 16]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
 
         (external_address, listen_address)
     }
+
+    /// * `key` - expects EXONUM_SERVICE_PASS or EXONUM_CONSENSUS_PASS.
+    fn get_passphrase(context: &Context, key: &str) -> Vec<u8> {
+        context
+            .get_flag_occurrences(NO_PASSWORD)
+            .map(|_| "".to_string())
+            .or_else(|| env::var(key).ok())
```
Comment (slowli) [inline exonum/src/helpers/fabric/details.rs]: Using a fixed env var can potentially leak a passphrase from another Exonum instance here, right? Doesn't look like a good decision to me.

File: exonum/src/helpers/fabric/details.rs [hunk 17]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
...
+            }).unwrap()
+            .into_bytes()
+    }
+
+    fn prompt_passphrase<R: BufRead, W: Write>(mut reader: R, mut writer: W) -> io::Result<String> {
+        loop {
+            write!(&mut writer, "Enter passphrase (empty for no passphrase): ")?;
+            let password = rpassword::read_password_with_reader(Some(&mut reader))?;
+            write!(&mut writer, "Enter same passphrase again: ")?;
+            if password == rpassword::read_password_with_reader(Some(&mut reader))? {
```
Comment (slowli) [inline exonum/src/helpers/fabric/details.rs]: I'd check that the password is non-empty, tbh. Empty passwords are acceptable for test networks, but otherwise an obvious security issue.

File: exonum/src/helpers/fabric/details.rs [hunk 18]
```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
...
+        passphrase: impl AsRef<[u8]>,
+    ) -> PublicKey {
+        let secret_key_path = secret_key_path.as_ref();
+        if secret_key_path.exists() {
+            panic!("{}: File exists", secret_key_path.to_string_lossy(),);
+        } else {
+            if let Some(dir) = secret_key_path.parent() {
+                fs::create_dir_all(dir).unwrap();
+            }
+            generate_keys_file(&secret_key_path, &passphrase).unwrap()
```
Comment (slowli) [inline exonum/src/helpers/fabric/details.rs]: Slightly off-topic: I'd be wary of using `unwrap()`s: they can leak diagnostic information useful to an adversary.

File: exonum/src/node/mod.rs [hunk 19]
```diff
@@ -268,6 +269,51 @@ pub struct NodeConfig {
...
+            .expect("Could not read service_secret_key from file")
+            .1;
+        NodeConfig {
+            consensus_secret_key,
+            service_secret_key,
+            genesis: self.genesis,
+            listen_address: self.listen_address,
+            external_address: self.external_address,
+            network: self.network,
+            consensus_public_key: self.consensus_public_key,
```
Comment (slowli) [inline exonum/src/node/mod.rs]: What happens if the keypair in the key file doesn't correspond to this public key?

File: exonum/src/helpers/fabric/builder.rs (hunk 14 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/builder.rs]: As far as I know, that's what on we agreed for the first iteration.

File: exonum/src/helpers/fabric/details.rs (hunk 15 above)
Comment (mvlabat) [reply exonum/src/helpers/fabric/details.rs]: Sounds reasonable... But it's not quite clear to me what kind of interface we should provide instead

File: exonum/src/node/mod.rs (hunk 19 above)
Comment (mvlabat) [reply exonum/src/node/mod.rs]: Nothing right now. Should we throw an error in this case?

I really discovered that we store public key in the key files when I published the PR. I think that another option is to remove public keys from our final config and read them from key files, so we can get rid of redundancy and duplication of data. What do you think about it?

File: exonum/src/node/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum/src/node/mod.rs]: Okay, sorry

File: exonum/src/helpers/fabric/details.rs [hunk 20]
```diff
@@ -416,6 +471,29 @@ impl Command for GenerateNodeConfig {
...
+                false,
+                "Path to the pem file storing service private key (default: ./service.toml)",
+                "s",
+                "service-path",
+                false,
+            ),
+            Argument::new_flag(
+                NO_PASSWORD,
+                "Don't prompt for passwords when generating private keys (leave empty)",
+                "n",
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum/src/helpers/fabric/details.rs]: Well, I think this functionality is potentially dangerous and every user should understand, that he's using "--no-password" flag, not some invisible "-n" :)

If you think that it's okay - I will not mind against.

File: exonum/src/node/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mvlabat) [reply exonum/src/node/mod.rs]: No problem :)

**Comment by dvush** [comment]
Moved to #1222

File: exonum/src/helpers/fabric/details.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dvush) [reply exonum/src/helpers/fabric/details.rs]: I've left it as is.

File: exonum/src/helpers/fabric/details.rs (hunk 3 above)
Comment (dvush) [reply exonum/src/helpers/fabric/details.rs]: I think since this function is used in "testnet" where there is no passphrase it can be left as is.
I changed scoping to address your comment in new PR.

File: exonum/src/helpers/fabric/details.rs (hunk 4 above)
Comment (dvush) [reply exonum/src/helpers/fabric/details.rs]: I used https://crates.io/crates/zeroize here.

File: exonum/src/helpers/fabric/details.rs (hunk 6 above)
Comment (dvush) [reply exonum/src/helpers/fabric/details.rs]: I thing that current arguments named better since function is called "path relative from" - first argument is "path" second argument "from".
Regarding moving is_absolute branch into some function I would consider premature since there are only two use cases, also this thing with paths may change(see comments later)

File: exonum/src/helpers/fabric/details.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (dvush) [reply exonum/src/helpers/fabric/details.rs]: It makes sence but for now I left it as is. Maybe UX will be slightly changed keeping this in mind.

File: exonum/src/helpers/fabric/details.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dvush) [reply exonum/src/helpers/fabric/details.rs]: I've added zeroizing Vec<u8> and string produced by read_password.

File: exonum/src/helpers/mod.rs (hunk 8 above)
Comment (dvush) [reply exonum/src/helpers/mod.rs]: This function does not work some of this cases properly.

>relative paths

Sort of works.

>special path components
>trying to escape the root directory

it does not "collapse" arguments so it does not work in all cases.

>("/", "/"), referring the directory itself

It will return ""...
